packages/core/custom-branding/core-custom-branding-browser-internal/src/custom_branding_service.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { BehaviorSubject, Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+import type { CustomBrandingStart, CustomBrandingSetup } from '@kbn/core-custom-branding-browser';
+import type { CustomBranding } from '@kbn/core-custom-branding-common';
+
+const CUSTOM_BRANDING_PLUGIN = 'customBranding';
+
+export class CustomBrandingService {
+  private customBranding: CustomBranding = {};
+  private registeredPlugin?: string;
+  private hasCustomBranding$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  private customBranding$: BehaviorSubject<CustomBranding> = new BehaviorSubject<CustomBranding>(
+    this.customBranding
+  );
+  private stop$ = new Subject<void>();
+
+  private set(customBranding: CustomBranding) {
+    if (this.registeredPlugin !== CUSTOM_BRANDING_PLUGIN) {
+      throw new Error('Plugin needs to register before setting custom branding');
+    }
+    Object.keys(customBranding).forEach((key) => {
+      this.customBranding[key as keyof CustomBranding] =
+        customBranding[key as keyof CustomBranding];
+    });
+    this.customBranding$.next(this.customBranding);
+    this.hasCustomBranding$.next(this.hasCustomBranding());
+  }
+
+  private hasCustomBranding() {
+    return Object.keys(this.customBranding).length > 0;
+  }
+
+  private register(pluginName: string) {
+    if (this.registeredPlugin) {
+      throw new Error('Plugin already registered');
+    }
+    this.registeredPlugin = pluginName;
+  }
+
+  /**
+   * @public
+   */
+  public start(): CustomBrandingStart {
+    return {
+      set: this.set.bind(this),
```

Comment:
Multiple questions here:

1. Why are we separating `register` and `set`?

I would expect something like an observable-based communication, registered during `setup`.

```ts
BrandingSetup {
   register(pluginOpaqueId: PluginId, customBranding$: Observable<CustomBranding>)
}

BrandingStart {
   customBranding$,
   hasCustomBranding$,
}
```

It seems so simpler and better in term of isolation of concern and separation of setup/start, wdyt?

2. Are we fine with the custom branding being propagated to Core from a plugin on the browser-side?

We talked about this a while ago, and the major downside I saw about this approach is that the customBranding plugin will be in charge of fetching the custom branding from the server.

Given the branding settings are now uiSettings, I'm assuming that your custom branding plugin will read the settings from the browser-side uisettings service and then inject them back to this new core custom branding service, is that correct? In that case, we don't have an additional roundtrip, so it's probably fine, but I'd like to have this confirmed.

Reply:
> Why are we separating register and set ?

If we already pass an Observable to the `register`, which service is then responsible to expose a `set` method to update the custom branding?